core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+                    nextEosTag                            = std::min(nextEosTag, samples_to_eos_tag(port).value_or(std::numeric_limits<std::size_t>::max()));
+                    const gr::ConsumableSpan auto tagData = port.tagReader().get(port.tagReader().available());
+                    hasTag |= !tagData.empty() && tagData[0].index == port.streamReader().position() && !tagData[0].map.empty();
+                } else { // async port
+                    if (samples_to_eos_tag(port).transform([&port](auto n) { return n <= port.min_samples; }).value_or(false)) {
+                        asyncEoS = true;
+                    }
+                }
+            }
+        };
+        for_each_port([&adjust_for_input_port](PortLike auto &port) { adjust_for_input_port(port); }, inputPorts<PortType::STREAM>(&self()));
+        return std::tuple{hasTag, nextTag, nextEosTag, asyncEoS};
+    }
+
+    /***
+     * skip leftover stride
+     * @param available number of samples that can be consumed from each sync port
+     * @return inputSamples to skip before the chunk
+     */
+    std::size_t
+    inputSamplesToSkipBeforeNextChunk(std::size_t available) {
+        std::size_t inputSamplesToSkipBefore = 0UZ;
+        if constexpr (StrideControl::kEnabled) {                    // check if stride was removed at compile time
+            if (stride.value != 0 && stride.value != denominator) { // check if stride is disabled at runtime
+                if (strideCounter != 0) {                           // drop samples
+                    if (strideCounter >= static_cast<gr::Size_t>(available)) {
+                        inputSamplesToSkipBefore = available;
+                        strideCounter -= static_cast<gr::Size_t>(available);
+                    } else {
+                        inputSamplesToSkipBefore = strideCounter;
+                        strideCounter            = 0;
+                    }
+                }
+            }
+        }
+        return inputSamplesToSkipBefore;
+    }
+
+    /***
+     * skip leftover stride
+     * @return inputSamples to skip before the chunk
+     */
+    std::size_t
+    inputSamplesToSkipAfterThisChunk(std::size_t remaining) {
+        // skip stride at the end
+        std::size_t inputSamplesToSkipAfter = 0UZ;
+        if constexpr (StrideControl::kEnabled) {
+            if (stride.value != 0 && stride.value != denominator) {
+                if (strideCounter == 0 && remaining > 0) {
+                    inputSamplesToSkipAfter = std::min(static_cast<std::size_t>(stride.value), remaining);
+                    strideCounter           = stride.value - static_cast<gr::Size_t>(inputSamplesToSkipAfter);
+                }
+            }
+        }
+        return inputSamplesToSkipAfter;
+    }
+
+    auto
+    decimate(std::size_t minSyncIn, std::size_t maxSyncIn, std::size_t minSyncOut, std::size_t maxSyncOut) {
```

Comment:
Considering that it can do both decimation and interpolation, wouldn't it be more appropriate to rename it to resample()? This also aligns with other names.